File: Engine/src/main/java/org/etieskrill/engine/entity/service/BoundingBoxRenderService.java
package org.etieskrill.engine.entity.service;

import org.etieskrill.engine.entity.Entity;
import org.etieskrill.engine.entity.component.Transform;
import org.etieskrill.engine.entity.component.WorldSpaceAABB;
import org.etieskrill.engine.graphics.Renderer;
import org.etieskrill.engine.graphics.camera.Camera;
import org.etieskrill.engine.graphics.gl.shader.Shaders.WireframeShader;
import org.etieskrill.engine.graphics.model.Model;
import org.etieskrill.engine.graphics.model.ModelFactory;
import org.joml.Vector3f;

import java.util.List;
import java.util.Set;

public class BoundingBoxRenderService implements Service {

    private final Renderer renderer;
    private final WireframeShader shader;
    private final Camera camera;

    private boolean renderBoundingBoxes;

    private Model box;
    private final Transform boundingBoxTransform = new Transform();

    public BoundingBoxRenderService(Renderer renderer, Camera camera) {
        this.renderer = renderer;
        this.shader = new WireframeShader();
        this.camera = camera;

        this.renderBoundingBoxes = true;
    }

    @Override
    public boolean canProcess(Entity entity) {
        return entity.hasComponents(WorldSpaceAABB.class);
    }

    @Override
    public void process(Entity targetEntity, List<Entity> entities, double delta) {
        if (!renderBoundingBoxes) return;
        renderBoundingBox(targetEntity.getComponent(WorldSpaceAABB.class));
    }

    private void renderBoundingBox(WorldSpaceAABB worldSpaceBoundingBox) {
        if (box == null) {
            box = ModelFactory.box(new Vector3f(1));
        }

        renderer.renderWireframe(
                boundingBoxTransform
                        .setPosition(worldSpaceBoundingBox.getCenter())
                        .setScale(worldSpaceBoundingBox.getSize()),
                box,
                shader,
                camera
        );
    }

    public void setRenderBoundingBoxes(boolean renderBoundingBoxes) {
        this.renderBoundingBoxes = renderBoundingBoxes;
    }

    public void toggleRenderBoundingBoxes() {
        renderBoundingBoxes = !renderBoundingBoxes;
    }

    @Override
    public Set<Class<? extends Service>> runAfter() {
        return Set.of(BoundingBoxService.class);
    }

}

File: Engine/src/jvmMain/java/org/etieskrill/engine/entity/service/impl/BoundingBoxRenderService.kt
package org.etieskrill.engine.entity.service.impl

import org.etieskrill.engine.entity.Entity
import org.etieskrill.engine.entity.component.Transform
import org.etieskrill.engine.entity.component.WorldSpaceAABB
import org.etieskrill.engine.entity.service.Service
import org.etieskrill.engine.graphics.Renderer
import org.etieskrill.engine.graphics.camera.Camera
import org.etieskrill.engine.graphics.gl.shader.Shaders
import org.etieskrill.engine.graphics.model.ModelFactory
import org.joml.Vector3f
import kotlin.reflect.KClass

class BoundingBoxRenderService(
    val renderer: Renderer,
    val camera: Camera
) : Service {

    var renderBoundingBoxes = true //TODO enable/disable services as a whole

    private val shader = Shaders.WireframeShader()

    private val box = ModelFactory.box(Vector3f(1f))
    private val boundingBoxTransform = Transform()

    override fun canProcess(entity: Entity) = entity.hasComponents<WorldSpaceAABB>()

    override fun process(targetEntity: Entity, entities: List<Entity>, delta: Double) {
        if (!renderBoundingBoxes) return

        val aabb = targetEntity.getComponent<WorldSpaceAABB>()!!
        aabb.center(boundingBoxTransform.position)
        aabb.getSize(boundingBoxTransform.scale)
        renderer.renderWireframe(boundingBoxTransform, box, shader, camera)
    }

    fun toggleRenderBoundingBoxes() {
        renderBoundingBoxes = !renderBoundingBoxes;
    }

    override val runAfter: Set<KClass<out Service>>
        get() = setOf(BoundingBoxService::class)

}
